### PR TITLE
Detect raid compound in root disk detection

### PR DESCRIPTION
**Pull request by schaefi**
Make sure raid managed storage e.g fake raid is detected
using the correct md device. Related to Issue #1181

### CODE REVIEW

**Comment by schaefi** [comment]
This is untested code and needs approval by someone with a fake raid controller at hand.

File: dracut/modules.d/99kiwi-lib/kiwi-lib.sh [hunk 1]
```diff
@@ -42,22 +42,37 @@ function lookup_disk_device_from_root {
...
+                fi
+            done
+        fi
+        # Check if root_device is managed by mdadm. If this
+        # is the case prefer the md device over the storage
+        # disks because this would create conflicts otherwise
+        if type mdadm &> /dev/null; then
+            for raid_disk in $(
+                mdadm --detail -Y "${disk_device}" 2>/dev/null |\
+                grep -E "MD_DEVICE_.*_DEV" | cut -f2 -d=
```
Comment (jpwidera) [inline dracut/modules.d/99kiwi-lib/kiwi-lib.sh]: Sincerly, on the Intel Rapid Storage device, the last entry in $disk_device is /dev/sda, which can't be handled by mdadm. Should be /dev/md*

Comment (schaefi) [reply dracut/modules.d/99kiwi-lib/kiwi-lib.sh]: mdadm --detail is called on any disk found. the ones that does not provide md information are not taken into account. The loop only runs if there is md information available and takes this md device and returns. I think if you read not by diff but by the complete code you will realized how things work

Comment (schaefi) [reply dracut/modules.d/99kiwi-lib/kiwi-lib.sh]: after all you should always have in mind this code must be generic enough for any disk situation including multipath, so we can't just mess it up or make it only for for raid-disks ;)

Comment (jpwidera) [reply dracut/modules.d/99kiwi-lib/kiwi-lib.sh]: Oh, excuse my rushing - I didn't see the do-loop changed.
Applying the code block gives at:
```
 mdadm --detail -Y "${disk_device}" 2>/dev/null |\
                grep -E "MD_DEVICE_.*_DEV" | cut -f2 -d=
```
`/dev/sda` and `/dev/sdb` as results.

Maybe it is easier to check, if 
```
mdadm --detail -Y "${disk_device}" 2>/dev/null
```
returns 0 (disk_device=/dev/sd* returns 1). In this case it is a valid md-disk.

Comment (schaefi) [reply dracut/modules.d/99kiwi-lib/kiwi-lib.sh]: yes you are right that's much easier. I've updated the code. Thanks

File: dracut/modules.d/99kiwi-lib/kiwi-lib.sh [hunk 2]
```diff
@@ -42,22 +41,31 @@ function lookup_disk_device_from_root {
...
+                    echo "${disk_device}"
+                    return
+                fi
+            done
+        fi
+        # Check if root_device is managed by mdadm. If this
+        # is the case prefer the md device over the storage
+        # disks because this would create conflicts otherwise
+        if type mdadm &> /dev/null; then
+            if mdadm --detail -Y "${disk_device}" 2>/dev/null; then
```
Comment (jpwidera) [inline dracut/modules.d/99kiwi-lib/kiwi-lib.sh]: This is good. Maybe a 
```
>/dev/null 2>&1
```
prevents the output from spoiling with unneeded information about the md-device (Additionally, the warning message, that any device isn't manageable by mdadm is stdout).

Comment (schaefi) [reply dracut/modules.d/99kiwi-lib/kiwi-lib.sh]: yep I changed it to &>/dev/null

**Comment by jpwidera** [comment]
@schaefi Thx, your fix does solve the problem (Tested on a fake-raid OEM-deployment).

**Comment by schaefi** [comment]
> I'll come back to this next week.

Thank you, much appreciated. I will merge this part next and submit a new package build to our staging area. You will find new packages at:

* https://build.opensuse.org/package/show/Virtualization:Appliances:Staging/python-kiwi

It would be great if you can build your image with the kiwi from that staging area and run the test with rd.kiwi.debug enabled

Again thanks for your help we wouldn't be there without you testing on real fakeraid devices

**Comment by jpwidera** [comment]
@schaefi 
I used the latest build on an empty directory. I can select and install a complete system into a fake-raid-device. And the fake-raid partitions are beeing resized.

Sadly, other problems came up with this (Not booting into kernel with kexec after deployment, and missing fake-raid-support in the deployed system) but **this** implementation works :)

> Once the system is up can you please send me the file /run/initramfs/log/boot.kiwi
> 
> I'd like to verify the partition resize of the raid. Thanks
> 
The system freezes after deployment, so I'm not able to get a log-file. I'll check this and go for a new issue

**Pull request merged**